### PR TITLE
[wip] Fix SNS max message size

### LINKED ISSUE #1776: We're exceeding SNS maximum message size
The id minter failed to send a message to SNS: ```com.amazonaws.services.sns.model.InvalidParameterException: Invalid parameter: Message too long (Service: AmazonSNS; Status Code: 400; Error Code: InvalidParameter; Request ID: 713d2d98-8ce1-5131-bd64-7ea6461406f5)```
It happened for bib record `b2043277x`, which does seem to have lots of items. As we add things to our transformations, we will exceed the limit earlier in the pipeline, so we need to decide how we address that.

- [x] Shared model
- [x] Pointer writer
- [x] Pointer reader

Move applications to use new messaging libraries:

- [x] `transformer` -> `id_minter` (#1944)
- [x] `id_minter` -> `ingestor`

### PR DESCRIPTION
**Pull request by rgladwell**
### What is this PR trying to achieve?

Fix #1776 and prevent AWS SNS/SQS messages exceeding maximum size limit.

### Have the following been considered/are they needed?

- [ ] Deployed new versions
- [ ] Run `terraform apply`.

### CODE REVIEW

File: common/src/main/scala/uk/ac/wellcome/sns/SNSWriter.scala [hunk 1]
```diff
@@ -6,32 +6,55 @@ import com.google.inject.Inject
...
-
-  private def toPublishRequest(message: String, subject: String) =
-    new PublishRequest(snsConfig.topicArn, message, subject)
+case class PublishAttempt(id: String, key: String)
+
+class SNSWriter @Inject()(
+  client: AmazonSNS,
+  config: SNSConfig,
+  s3Client: AmazonS3,
+  s3Config: S3Config
```
Comment (alicefuzier) [inline common/src/main/scala/uk/ac/wellcome/sns/SNSWriter.scala]: I think the design that we came up with is not really to modify `SNSWriter` to write to s3 but rather to write a wrapper object (we haven't really decided on a name yet, something like `SNSPointerWriter`) that uses `SNSWriter` to send to a SNS message and `S3ObjectStore` to write to s3. So `SNSWriter` probably shouldn't change much

Comment (alicefuzier) [reply common/src/main/scala/uk/ac/wellcome/sns/SNSWriter.scala]: Paging @alexwlchan and @kenoir just to make sure we're all on the same page

Comment (alexwlchan) [reply common/src/main/scala/uk/ac/wellcome/sns/SNSWriter.scala]: Yeah, this looks wrong to me. I was thinking the existing SNSWriter would stay the same, and we'd add an entirely new class.

File: common/src/test/scala/uk/ac/wellcome/sns/SNSWriterTest.scala [hunk 2]
```diff
@@ -3,49 +3,88 @@ package uk.ac.wellcome.sns
...
+
+          getContentFromS3(bucket, publishAttempt.right.value.key) should be(
+            message)
+        }
+      }
+    }
+  }
+
+  it(
+    "should return a failed future if it fails to publish the message pointer") {
```
Comment (kenoir) [inline common/src/test/scala/uk/ac/wellcome/sns/SNSWriterTest.scala]: We try and avoid `should` in test descriptions as it's redundant. The test asserts something _will_ happen and fails if it doesn't. Apologies for the pedantry.

Comment (kenoir) [reply common/src/test/scala/uk/ac/wellcome/sns/SNSWriterTest.scala]: This set of tests looks good though.

File: catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/receive/SQSMessageReceiverTest.scala [hunk 3]
```diff
@@ -72,7 +72,12 @@ class SQSMessageReceiverTest
   )(testWith: TestWith[SQSMessageReceiver, R]) = {
 
     val snsWriter =
-      maybeSnsWriter.getOrElse(new SNSWriter(snsClient, SNSConfig(topic.arn)))
+      maybeSnsWriter.getOrElse(
+        new SNSWriter(
+          snsClient,
+          SNSConfig(topic.arn),
+          s3Client,
```
Comment (kenoir) [inline catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/receive/SQSMessageReceiverTest.scala]: You can probably just inject: https://github.com/wellcometrust/platform/blob/master/common/src/main/scala/uk/ac/wellcome/s3/S3ObjectStore.scala

Comment (kenoir) [reply catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/receive/SQSMessageReceiverTest.scala]: 👍 Presumed this was implicit from previous comments.

File: common/src/main/scala/uk/ac/wellcome/s3/S3Uri.scala [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package uk.ac.wellcome.s3
+
+import java.net.URI
+
+object S3Uri {
```
Comment (kenoir) [inline common/src/main/scala/uk/ac/wellcome/s3/S3Uri.scala]: Why use a URI over having a mechanism to serialise/deserialise: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/S3Object.html?

Comment (kenoir) [reply common/src/main/scala/uk/ac/wellcome/s3/S3Uri.scala]: There's also https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3URI.html that more conveniently allows us to extract bucket/key from a URI.

Comment (kenoir) [reply common/src/main/scala/uk/ac/wellcome/s3/S3Uri.scala]: Perhaps actually using a plain URI here is ok - as we can use `AmazonS3URI` the other side to extract the bucket and key, making the sending side know a bit less about the storage medium.

**Pull request closed without merging**